### PR TITLE
Adds a SecretDict field that obfuscated values in a dictionary

### LINKED ISSUE #6735: Add a SecretJson / SecretDict type similar to SecretStr in Pydantic
### First check

- [X] I added a descriptive title to this issue.
- [X] I used the GitHub search to find a similar request and didn't find it.
- [X] I searched the Prefect documentation for this feature.

### Prefect Version

2.x

### Describe the current behavior

Currently, we have this implementation in prefect-gcp:
```
class GcpCredentials(Block):
    ...
    service_account_info: Optional[Union[Dict[str, str], Json]] = None
```

Logging/printing the credentials shows the entire secret.

Han also asked about this when he was developing prefect-fugue:
https://github.com/fugue-project/prefect-fugue/blob/master/prefect_fugue/blocks.py#L43-L47

Although it's a SecretStr, it still shows in the UI
image

### Describe the proposed behavior

Ideally:
```
class GcpCredentials(Block):
    ...
    service_account_info: Optional[Union[SecretDict[str, str], SecretJson]] = None
```
Or:
```
    secret_conf: Optional[SecretDict[str, str]] = Field(
        default=None,
        alias="secret_config",
        description="A JSON-dict-compatible value",
    )
```
And the secret values would be obscured.

### Example Use

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by desertaxle**
Adds a `SecretDict` field that can be used within blocks to obfuscate the values in a dictionary. This is useful for blocks like the [`FugueEngine` block](https://fugue-project.github.io/prefect-fugue/blocks/#prefect_fugue.blocks.FugueEngine) which can accept a wide variety of credentials which are not known until configuration or the [`GcpCredentials` block](https://prefecthq.github.io/prefect-gcp/credentials/#prefect_gcp.credentials.GcpCredentials) which sometimes requires sensitive information in a JSON format.

To implement the obfuscation in the UI, the `secret_fields` parameter of a block schema is extended to allow a wildcard character within a secret field path (e.g. `service_account_info.*`). If a path with a wildcard is found in a block schema's `secret_fields`, then the values for all keys nested in that path will be obfuscated by the UI. When logging a `SecretDict` value, only the top level keys are shown and all values and nested keys are hidden. When generating a block schema for a block with a `SecretDict` field, the path of the `SecretDict` field with a `.*` appended at the end will be added to `secret_fields`, so the user only needs to type a block attribute with `SecretDict` to receive these benefits.

Closes #6735 

### Example

Declare a block that has a `SecretDict` field:
```python
from prefect.blocks.core import Block
from prefect.blocks.fields import SecretDict

class BoxOfSecrets(Block):
    diary_entries: SecretDict
```

Create an instance of a block with a `SecretDict` field:
```python
my_box_of_secrets = BoxOfSecrets(
    diary_entries={
       "12-13-2022": "Accidentally wore my shirt inside out to first period.",
       "12-14-2022": "Fell in a puddle today. It was super embarrassing!"
    }
)
```

Values are obfuscated when printed or logged:
```python
print(my_box_of_secrets)
```
Output:
```bash
BoxOfSecrets(diary_entries=SecretDict('{'12-13-2022': '**********', '12-14-2022': '**********'}'))
```

Values are obfuscated when displayed in the UI:
Screenshot 2022-12-14 at 7 14 16 AM

Values are also obfuscated when editing a block in the UI:
Screenshot 2022-12-14 at 7 16 41 AM

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
	- If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] This pull request includes tests or only affects documentation.
- [x] This pull request includes a label categorizing the change e.g. `fix`, `feature`, `enhancement`

### CODE REVIEW

File: src/prefect/blocks/core.py [hunk 1]
```diff
@@ -41,6 +41,8 @@
 from prefect.utilities.importtools import to_qualified_name
 from prefect.utilities.slugify import slugify
 
+from .fields import SecretDict
```
Comment (ahuang11) [inline src/prefect/blocks/core.py:44]:
```suggestion
from prefect.blocks.fields import SecretDict
```

File: src/prefect/blocks/core.py [hunk 2]
```diff
@@ -143,6 +145,8 @@ def schema_extra(schema: Dict[str, Any], model: Type["Block"]):
             for field in model.__fields__.values():
                 if field.type_ in [SecretStr, SecretBytes]:
                     secrets.append(field.name)
+                if field.type_ == SecretDict:
```
Comment (ahuang11) [inline src/prefect/blocks/core.py:148]:
```suggestion
                elif field.type_ == SecretDict:
```

File: src/prefect/blocks/core.py [hunk 3]
```diff
@@ -647,6 +651,7 @@ class Custom(Block):
             block_document = await client.read_block_document_by_name(
                 name=block_document_name, block_type_slug=block_type_slug
             )
+            print(block_document)
```
Comment (ahuang11) [inline src/prefect/blocks/core.py:654]:
```suggestion
```

File: src/prefect/orion/schemas/core.py [hunk 4]
```diff
@@ -663,12 +662,17 @@ async def from_orm_model(
...
                 "secret_fields", []
             ):
-                key = tuple(field.split("."))
-                if flat_data.get(key) is not None:
-                    flat_data[key] = obfuscate_string(flat_data[key])
+                secret_key = tuple(secret_field.split("."))
+                if flat_data.get(secret_key) is not None:
+                    flat_data[secret_key] = obfuscate_string(flat_data[secret_key])
+                elif "*" in secret_key:
+                    wildcard_index = secret_key.index("*")
```
Comment (ahuang11) [inline src/prefect/orion/schemas/core.py:672]: Not sure if this is necessary, but what if we use `.*` to narrow it down:
```suggestion
                elif ".*" in secret_key:
                    wildcard_index = secret_key.index(".*")
```

Comment (ahuang11) [reply src/prefect/orion/schemas/core.py:672]: Also, might be nice to leave some comments on this.

**Review by ahuang11 — COMMENTED** [review]
Some draft comments.

File: src/prefect/orion/schemas/core.py (hunk 4 above)
Comment (desertaxle) [reply src/prefect/orion/schemas/core.py:676]: `secret_key` is a tuple created by splitting the field path on periods, so this wouldn't work. I've added some comments to shed some light on how the obfuscation in the backend works.

**Review by tpdorsey — APPROVED** [review]
LGTM

File: docs/concepts/blocks.md [hunk 5]
```diff
@@ -150,6 +150,40 @@ print(aws_credentials_block)
...
+        "self_destruct_countdown_seconds": 60,
+        "self_destruct_countdown_stop_time": 7,
+    },
+)
+```
+`system_secrets` will be obfuscated if `system_configuration_block` is logged, but `system_variables` will be shown in plain-text:
+
+```python
+print(system_configuration_block)
+# SystemConfiguration(system_secrets=SecretDict('{'password': '**********', 'api_token': '**********', 'private_key': '**********'}'), system_variables={'self_destruct_countdown_seconds': 60, 'self_destruct_countdown_stop_time': 7})
```
Comment (madkinsz) [inline docs/concepts/blocks.md:185]: fwiw I'd use `pprint`

File: docs/concepts/blocks.md [hunk 6]
```diff
@@ -150,6 +150,40 @@ print(aws_credentials_block)
...
+        "api_token": "token_123456789",
+        "private_key": "<private key here>",
+    },
+    system_variables={
+        "self_destruct_countdown_seconds": 60,
+        "self_destruct_countdown_stop_time": 7,
+    },
+)
+```
+`system_secrets` will be obfuscated if `system_configuration_block` is logged, but `system_variables` will be shown in plain-text:
```
Comment (madkinsz) [inline docs/concepts/blocks.md:181]:
```suggestion
`system_secrets` will be obfuscated when `system_configuration_block` is displayed, but `system_variables` will be shown in plain-text:
```

File: docs/concepts/blocks.md [hunk 7]
```diff
@@ -150,6 +150,40 @@ print(aws_credentials_block)
 # aws_access_key_id='AKIAJKLJKLJKLJKLJKLJK' aws_secret_access_key=SecretStr('**********') aws_session_token=None profile_name=None region_name=None
 ```
 
+For more complex secret values, you can also use the `SecretDict` field type provided by Prefect. This type will allow you to add a dictionary field to your block that will have values at all levels automatically obfuscated in the UI or in logs. This is useful for blocks where typing or structure of secret fields is not known until configuration time.
```
Comment (ahuang11) [inline docs/concepts/blocks.md:153]:
```suggestion
There's  also use the `SecretDict` field type provided by Prefect. This type will allow you to add a dictionary field to your block that will have values at all levels automatically obfuscated in the UI or in logs. This is useful for blocks where typing or structure of secret fields is not known until configuration time.
```

File: src/prefect/blocks/core.py [hunk 8]
```diff
@@ -120,6 +121,8 @@ class Block(BaseModel, ABC):
     class Config:
         extra = "allow"
 
+        json_encoders = {SecretDict: lambda v: v.dict()}
```
Comment (ahuang11) [inline src/prefect/blocks/core.py:124]: Do we need to get_secret_values here since the value gets irreversibly obfuscated?

File: src/prefect/orion/models/block_documents.py [hunk 9]
```diff
@@ -445,14 +446,28 @@ async def update_block_document(
...
+            # and they are replaced with the current value for that key to avoid losing
+            # data during update.
+            elif "*" in secret_key:
+                wildcard_index = secret_key.index("*")
+                for data_key in flat_update_data.keys():
+                    if secret_key[0:wildcard_index] == data_key[
+                        0:wildcard_index
+                    ] and flat_update_data[data_key] == obfuscate(
+                        flat_update_data[data_key]
+                    ):
```
Comment (madkinsz) [inline src/prefect/orion/models/block_documents.py:468]: Black formats these so weird. Generally, adding parens will make it more readable.

File: src/prefect/blocks/core.py (hunk 8 above)
Comment (desertaxle) [reply src/prefect/blocks/core.py:124]: No, this is to make sure that you can't see the values when `.json()` is called on a block the values aren't printed. Getting the secret values to send to the API happens when we create a block document with `_to_block_document`

Comment (madkinsz) [reply src/prefect/blocks/core.py:124]: Yeah this seems like a JSON roundtrip would fail?

File: src/prefect/orion/models/block_documents.py [hunk 10]
```diff
@@ -532,10 +549,6 @@ async def create_block_document_reference(
     block_document_reference: schemas.actions.BlockDocumentReferenceCreate,
     db: OrionDBInterface,
 ):
-    values = block_document_reference.dict(
-        shallow=True, exclude_unset=True, exclude={"created", "updated"}
-    )
-
```
Comment (madkinsz) [inline src/prefect/orion/models/block_documents.py:538]: Why is this removed? Unused?

File: docs/concepts/blocks.md (hunk 5 above)
Comment (desertaxle) [reply docs/concepts/blocks.md:185]: I tried using `pprint` just now and it's still not pretty

File: src/prefect/orion/models/block_documents.py (hunk 10 above)
Comment (desertaxle) [reply src/prefect/orion/models/block_documents.py:538]: Yeah it was an used variable.

File: src/prefect/orion/utilities/schemas.py [hunk 11]
```diff
@@ -157,6 +157,10 @@ class Config:
         else:
             extra = "ignore"
 
+        json_encoders = {
+            SecretField: lambda v: v.dict() if getattr(v, "dict", None) else str(v)
+        }
```
Comment (madkinsz) [inline src/prefect/orion/utilities/schemas.py:162]: This feels... weird. Can we handle `SecretDict` separately?

File: docs/concepts/blocks.md (hunk 5 above)
Comment (madkinsz) [reply docs/concepts/blocks.md:185]: Oh well, I often will add manual line breaks and indentation so it's readable in the doc even if it's not the real output

File: src/prefect/orion/utilities/schemas.py (hunk 11 above)
Comment (desertaxle) [reply src/prefect/orion/utilities/schemas.py:162]: I did this to avoid a circular import. Does it make sense to include `SecretDict` in this file?

File: src/prefect/blocks/core.py (hunk 8 above)
Comment (madkinsz) [reply src/prefect/blocks/core.py:124]: I feel like we should only obfuscate values in `str` and `repr` but 🤷‍♀️

File: src/prefect/orion/utilities/schemas.py (hunk 11 above)
Comment (madkinsz) [reply src/prefect/orion/utilities/schemas.py:162]: Ah yeesh. Just note it in a comment and we can tackle it whenever I declare war on circular imports.

File: src/prefect/blocks/core.py (hunk 8 above)
Comment (desertaxle) [reply src/prefect/blocks/core.py:124]: The JSON payload comes from the `BlockDocument` model and not from the `Block` subclass directly. It might be too aggressive to obfuscate in `.json()` as well, but I thought it was the most expected behavior.

Comment (madkinsz) [reply src/prefect/blocks/core.py:124]: Let's revisit this if/when it's a problem :)

**Review by madkinsz — APPROVED** [review]
